mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java

```diff
@@ -0,0 +1,191 @@
+
+/** Provides mechanisms for building and traversing a class/interfaces hierarchy. */
+public class HierarchyScanner {
+
+  // placeholder class for resolving the default HierarchyScanStrategy
+  @HierarchyScanStrategy
+  private static final class ClassForDefaultScanStrategy {}
+
+  /**
+   * Resolves the type hierarchy for the given type by first looking for a {@link
+   * HierarchyScanStrategy}-annotated class or interface in the given types hierarchy. The hierarchy
+   * is traversed until a {@link HierarchyScanStrategy} is encountered in a depth-first manner as
+   * follows:
+   *
+   * <ol>
+   *   <li>Initialize <code>i</code> as an empty set
+   *   <li>Visit typeElement (<code>e</code>)
+   *   <li>Visit <code>e</code>'s parent class (<code>p</code>), if <code>p</code> is null, stop
+   *       traversing after evaluating remaining interfaces
+   *   <li>Visit <code>e</code>'s interfaces, and record those interfaces' parents for later use (
+   *       <code>ni</code>)
+   *   <li>Visit <code>i</code>, and append those interface's parents to <code>ni</code> for later
+   *       use)
+   *   <li>If <code>p != null</code> </code>Set <code>e := p, i := ni</code> and repeat starting at
+   *       step 3
+   *   <li>Visit <code>ni</code> interfaces and their parents until we've reached root
+   * </ol>
+   *
+   * Once a {@link HierarchyScanStrategy} is identified, the returning hierarchy is built by
+   * traversing the hierarchy again using the chosen strategy.
+   *
+   * @param typeElement The type whose hierarchy will be traversed.
+   * @param types Used to resolve parent type elements
+   * @return The type hierarchy, ordered from typeElement to the highestAncestor, as dictated by the
+   *     resolved {@link HierarchyScanStrategy}
+   */
+  public static Set<TypeElement> resolveTypeHierarchy(TypeElement typeElement, Types types) {
+    HierarchyScanStrategy hierarchyScanStrategy =
+        HierarchyScanner.resolveHierarchyScanStrategy(typeElement, types);
+
+    ImmutableSet.Builder<TypeElement> hierarchy = ImmutableSet.builder();
+    traverseFullHierarchy(hierarchyScanStrategy, typeElement, types, hierarchy::add);
+    return hierarchy.build();
+  }
+
+  private static HierarchyScanStrategy resolveHierarchyScanStrategy(
+      TypeElement classElement, Types types) {
+    // Use the default HierarchyScanStrategy to find the configured HierarchyScanStrategy.
+    // This is done because the default strategy is the most permissive.
+    HierarchyScanStrategy strategy =
+        ClassForDefaultScanStrategy.class.getAnnotation(HierarchyScanStrategy.class);
+
+    // traverse hierarchy until a strategy is found.
+    final AtomicReference<HierarchyScanStrategy> ref = new AtomicReference<>(strategy);
+    traverseHierarchy(
+        strategy,
+        classElement,
+        types,
+        (TypeElement t) -> {
+          HierarchyScanStrategy discoveredStrategy = t.getAnnotation(HierarchyScanStrategy.class);
```

Comment:
I'm surprised that this works. Generally you can't inspect an annotation's `Class<?>` attribute during a processing round, because it might not be possible to load the referenced class yet (see `DefaultEntityFactory#readCustomConverterClasses` for example).

Have you tried with `HierarchyScanStrategy#highestAncestor()` being an application class?

Reply:
I'm not confident it will work as I haven't tested it yet at the compile phase, will see how it goes today ;)

Reply:
Your right, it doesn't work, will follow the strategy you referenced :+1:

```
Error:(676, 3) java: Unexpected error while writing generated code: javax.lang.model.type.MirroredTypeException: Attempt to access Class object for TypeMirror com.datastax.oss.driver.mapper.EntityPolymorphismIT.Asset
...
com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.traverseHierarchy(HierarchyScanner.java:136)
  	at com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.traverseFullHierarchy(HierarchyScanner.java:111)
  	at com.datastax.oss.driver.internal.mapper.processor.util.HierarchyScanner.resolveTypeHierarchy(HierarchyScanner.java:75)
```

Reply:
Fixed this in 8fdcc880a1f6050cf7b915127e29f5c4c1a56c25